### PR TITLE
Pipeline2.0 STEP2 - move hybrid buffering logic from comp_buffer to audio_buffer and partially place comp_buffer on top of audio_buffer 

### PR DESCRIPTION
**Pull request by marcinszkudlinski**
1) all logic of hybrid buffering - double buffering, currently comp_buffer <=> ring_buffer, moved to audio_buffer layer
2) comp_buffer initially uses audio_buffer API - sink/source removed from audio_stream

next steps would be to remove all access to comp_buffer except by 3APIs: audio_buffer / sink / source
Those APIs may of course get another functions during process

### CODE REVIEW

**Comment by marcinszkudlinski** [comment]
@lgirdwood the flag is enabled for IPC4 platforms. 

- Some features, like DP processing, require it
- The changes in this PR are transparent for functionality
- for legacy platforms - disabling the flag is just saving memory

**Review by kv2019i — COMMENTED** [review]
Code looks good, ok to +1 once CI is clean. Some build warnings that need to be addressed like https://github.com/thesofproject/sof/actions/runs/10316722150/job/28559534467?pr=9299

**Comment by marcinszkudlinski** [comment]
fixin compilation in progress, there are still some work to do so DNM

**Comment by marcinszkudlinski** [comment]
memory leak fixed
compilation problems fixed
please proceed

**Comment by marc-hb** [comment]
The input/output error in  https://sof-ci.01.org/sofpr/PR9299/build7167/devicetest/index.html?model=LNLM_SDW_AIOC&testcase=check-capture-3times
seems completely new.

```
arecord: pcm_read:2221: read error: Input/output error
2024-08-20 11:24:54 UTC [REMOTE_ERROR] arecord on PCM hw:0,3 failed at 1/3.
2024-08-20 11:24:54 UTC [REMOTE_INFO] Starting func_exit_handler(1)
2024-08-20 11:24:54 UTC [REMOTE_ERROR] Starting func_exit_handler(), exit status=1, FUNCNAME stack:
2024-08-20 11:24:54 UTC [REMOTE_ERROR]  die()  @  /home/ubuntu/sof-test/test-case/../case-lib/lib.sh
2024-08-20 11:24:54 UTC [REMOTE_ERROR]  main()  @  /home/ubuntu/sof-test/test-case/check-capture.sh:101

````

**Comment by marc-hb** [comment]
SOFCI TEST

**Comment by marcinszkudlinski** [comment]
@lgirdwood @kv2019i can we proceed?

**Review by kv2019i — APPROVED** [review]

**Comment by kv2019i** [comment]
@marcinszkudlinski Can you check Intel internal/quickbuild fail? That's the only thing holding back merge.

**Comment by marcinszkudlinski** [comment]
@lgirdwood looks like there was a problem with FPGA testing, hopefully fixed
I'll rebase on head and will try again

File: src/audio/CMakeLists.txt [hunk 1]
```diff
@@ -128,7 +129,7 @@ endif()
 add_local_sources(sof
 	component.c
 	data_blob.c
-	buffer.c
+	buffers/comp_buffer.c
```
Comment (lgirdwood) [inline src/audio/CMakeLists.txt:132]: We can do this in the next PR if needed (as CI is good), but does it make sense to rename like other similar files to legacy-buffer.c ?

**Review by lgirdwood — APPROVED** [review]

File: src/audio/buffers/audio_buffer.c [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+					 struct sof_audio_buffer *secondary_buffer)
+{
+	if (buffer->secondary_buffer_sink || buffer->secondary_buffer_source)
+		return -EINVAL;
+
+	/* secondary buffer must share audio params with the primary buffer */
+	secondary_buffer->audio_stream_params = buffer->audio_stream_params;
+	/* for performance reasons pointers to params are also kept in sink/src structures */
+	secondary_buffer->_sink_api.audio_stream_params = buffer->audio_stream_params;
+	secondary_buffer->_source_api.audio_stream_params = buffer->audio_stream_params;
```
Comment (lyakh) [inline src/audio/buffers/audio_buffer.c:28]: this seems to make 3 copies of stream parameters, so now we have 4 of them on the total? This seems really redundant. Is this a temporary situation until buffers are migrated to the new API? TBH I'm a bit confused about the roadmap of the migration, how far we're into it and what is still left to do. @marcinszkudlinski maybe you could make a "feature" to explain the steps made so far, the current configuration (what types of buffers we currently have, what copying is done, etc.) and the roadmap?

**Comment by marc-hb** [comment]
> @marcinszkudlinski maybe you could make a "feature" to explain the steps made so far, the current configuration (what types of buffers we currently have, what copying is done, etc.) and the roadmap?

Maybe like this? https://github.com/orgs/thesofproject/discussions/8645

I assumed this link was already there somewhere in the commit messages or PR description or source code comments but maybe it wasn't.

**Comment by marcinszkudlinski** [comment]
@lyakh @marc-hb 
Yes, I need to update the doc you mention 
I'm doing pipeline2,0 in a "desing&implement" mode, so archeo drifts also a bit when new issues&possibilites come into my mind. 

Also, I roadmap would be handy.

**Comment by lyakh** [comment]
> and readable version https://marcinszkudlinski.github.io/sof-docs/PAGES/architectures/firmware/sof-common/pipeline_2_0/pipeline2_0_discussion.html

@marcinszkudlinski thanks, but is that outdated - I don't see it mentioning `struct sof_audio_buffer` at all? So now we have 3 APIs: the "original / stable / legacy" `struct comp_buffer,` source / sink and the new `struct sof_audio_buffer`? So would be good to know where we are now - which APIs are used by which components and how many additional copies get done when converting between APIs?

**Pull request merged**